# THIS IS A CONTENT BASED MOVIE RECOMMENDER SYSTEM

In [2]:
# from google.colab import drive
# drive.mount('/content/drive')

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
mov=pd.read_csv('tmdb_5000_movies.csv')
cred=pd.read_csv('tmdb_5000_credits.csv')

In [5]:
mov.head(5)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

In [6]:
cred.head(5)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [7]:
mov.shape

(4803, 20)

In [8]:
cred.shape

(4803, 4)

### MERGING BOTH THE DATAFRAMES INTO ONE DATAFRAME

In [9]:
all_df=mov.merge(cred,on='title')

In [10]:
all_df.shape

(4809, 23)

In [11]:
all_df.head(5)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  ... runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   169.0   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   148.0   
3  [{"name": "Legendary Pictures", "id": 923}, {"...  ...   165.0   
4        [{"name": "Walt Disney Pictures", "id": 2}]  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count movie_id  \
0                                    Avatar          7.2      11800    19995   
1  Pirates of the Caribbean: At World's End          6.9       4500      285   
2                                   Spectre          6.3       4466   206647   
3                     The Dark Knight Rises          7.6       9106    49026   
4                               John Carter          6.1       2124    49529   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [12]:
all_df.isnull().sum()

budget                     0
genres                     0
homepage                3096
id                         0
keywords                   0
original_language          0
original_title             0
overview                   3
popularity                 0
production_companies       0
production_countries       0
release_date               1
revenue                    0
runtime                    2
spoken_languages           0
status                     0
tagline                  844
title                      0
vote_average               0
vote_count                 0
movie_id                   0
cast                       0
crew                       0
dtype: int64

## As I'm building a Content-based Movie Recommender System, I am including only columns or variables, which have categorical data.

In [13]:
all_df=all_df[['id','title','genres','overview','keywords','cast','crew']]
all_df.head(3)

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...

In [14]:
all_df.shape

(4809, 7)

In [15]:
all_df.isnull().sum()

id          0
title       0
genres      0
overview    3
keywords    0
cast        0
crew        0
dtype: int64

In [16]:
np.round(all_df.isnull().mean()*100,2) # it gives the percentage of null values

id          0.00
title       0.00
genres      0.00
overview    0.06
keywords    0.00
cast        0.00
crew        0.00
dtype: float64

In [17]:
all_df.dropna(inplace=True)

In [18]:
np.round(all_df.isnull().mean()*100,2)

id          0.0
title       0.0
genres      0.0
overview    0.0
keywords    0.0
cast        0.0
crew        0.0
dtype: float64

In [19]:
all_df.duplicated().sum() # This implies there are no duplicated records or rows.

0

In [20]:
all_df['genres'][0]

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

### AS all_df['genres'] CONTAINS STRINGS in which there is a LIST OF DICTIONARIES WHICH CONTAINS GENRES, WE NEED TO CONVERT IT INTO A LIST FIRST

In [21]:
import ast

### Using "ast" MODULE FOR COVERTING STRING (in which it contains lists) TO LISTS, BY REMOVING THE STRING

In [22]:
ast.literal_eval(all_df['genres'][0]) # IT WORKS LIKE THIS

[{'id': 28, 'name': 'Action'},
 {'id': 12, 'name': 'Adventure'},
 {'id': 14, 'name': 'Fantasy'},
 {'id': 878, 'name': 'Science Fiction'}]

In [23]:
# Function for converting strings(in which it contains lists) into lists
def release(obj):
    l=[]
    for i in ast.literal_eval(obj):
       l.append(i['name'])
    return l

In [24]:
all_df['genres']=all_df['genres'].apply(release) # applting function to all_df['genres']

In [25]:
all_df.sample(3)

id                            title            genres  \
4239   14054                      Snow Angels  [Drama, Romance]   
3950  317930                        Compadres  [Action, Comedy]   
3726     510  One Flew Over the Cuckoo's Nest           [Drama]   

                                               overview  \
4239  Waitress Annie has separated from her suicidal...   
3950  A disgruntled Mexican cop is forced to work wi...   
3726  While serving time for insanity at a state men...   

                                               keywords  \
4239  [{"id": 236, "name": "suicide"}, {"id": 894, "...   
3950                                                 []   
3726  [{"id": 30, "name": "individual"}, {"id": 526,...   

                                                   cast  \
4239  [{"cast_id": 2, "character": "Annie Marchand",...   
3950  [{"cast_id": 3, "character": "", "credit_id": ...   
3726  [{"cast_id": 15, "character": "Randle Patrick ...   

                                                   crew  
4239  [{"credit_id": "52fe45c79251416c750622ef", "de...  
3950  [{"credit_id": "54b5c112c3a36838d500034b", "de...  
3726  [{"credit_id": "52fe424cc3a36847f80134e7", "de...

In [26]:
all_df['keywords'][0] # it is in same format as in "genres" variable

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

In [27]:
all_df['keywords']=all_df['keywords'].apply(release) # So, Applying "release" function to all_df['keywords']
all_df.sample(3)

id               title                     genres  \
1118   11812  The Sweetest Thing          [Romance, Comedy]   
2060  109513           Hit & Run  [Action, Comedy, Romance]   
4344   11298         The Howling            [Drama, Horror]   

                                               overview  \
1118  Christina's love life is stuck in neutral. Aft...   
2060  Former getaway driver Charlie Bronson jeopardi...   
4344  After a bizarre and near fatal encounter with ...   

                                               keywords  \
1118  [roommate, marriage, aftercreditsstinger, duri...   
2060  [witness protection, getaway driver, duringcre...   
4344               [insane asylum, werewolf, newswoman]   

                                                   cast  \
1118  [{"cast_id": 10, "character": "Christina Walte...   
2060  [{"cast_id": 11, "character": "Annie Bean", "c...   
4344  [{"cast_id": 10, "character": "Dr. George Wagg...   

                                                   crew  
1118  [{"credit_id": "52fe448a9251416c7503857f", "de...  
2060  [{"credit_id": "595b9484925141222b039eac", "de...  
4344  [{"credit_id": "52fe44209251416c75029ad1", "de...

In [28]:
all_df['cast'][0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

<h3>As we have enormous number of cast members in every record, I am just going with top three(3) casts i.e.</h3>
<p>-> {"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}</p>
<p>-> {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}</p>
<p>-> {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}</p>

In [29]:
# Function for converting strings(in which it contains lists) into lists
# (This function also selects top 3[three] cast names in the movie.)
def releaseTop3(obj):
    l=[]
    c=0
    for i in ast.literal_eval(obj):
        if c<=2:
            l.append(i['name'])
            c+=1
        else:
            break
    return l

In [30]:
all_df['cast']=all_df['cast'].apply(releaseTop3)

In [31]:
all_df.sample(3)

id              title                    genres  \
4492  302579   Naturally Native                   [Drama]   
3812   20065     Purple Violets  [Comedy, Drama, Romance]   
4808   25975  My Date with Drew             [Documentary]   

                                               overview  \
4492  Naturally Native follows the lives, loves, pai...   
3812  Patti Petalson is a promising writer, but her ...   
4808  Ever since the second grade when he first saw ...   

                                               keywords  \
4492  [native american, racism, family, woman director]   
3812                                 [independent film]   
4808          [obsession, camcorder, crush, dream girl]   

                                                   cast  \
4492    [Valerie Red-Horse, Yvonne Russo, Irene Bedard]   
3812        [Selma Blair, Patrick Wilson, Edward Burns]   
4808  [Drew Barrymore, Brian Herzlinger, Corey Feldman]   

                                                   crew  
4492  [{"credit_id": "5462d90c0e0a2677fd00116d", "de...  
3812  [{"credit_id": "52fe43d5c3a368484e000cbb", "de...  
4808  [{"credit_id": "58ce021b9251415a390165d9", "de...

In [32]:
all_df['crew'][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

### As we have enormous number of crew members in every record, I am just going with Director

In [33]:
# Function for converting strings(in which it contains lists) into lists and for Selecting movie director's name.
def release_director_name(obj):
    l=[]
    for i in ast.literal_eval(obj):
        if i['job']=='Director':
            l.append(i['name'])
    return l

In [34]:
all_df['crew'].apply(release_director_name)

0                                [James Cameron]
1                               [Gore Verbinski]
2                                   [Sam Mendes]
3                            [Christopher Nolan]
4                               [Andrew Stanton]
                          ...                   
4804                          [Robert Rodriguez]
4805                              [Edward Burns]
4806                               [Scott Smith]
4807                               [Daniel Hsia]
4808    [Brian Herzlinger, Jon Gunn, Brett Winn]
Name: crew, Length: 4806, dtype: object

In [35]:
all_df['crew'][4808]

'[{"credit_id": "58ce021b9251415a390165d9", "department": "Production", "gender": 2, "id": 6888, "job": "Executive Producer", "name": "Clark Peterson"}, {"credit_id": "58ce0232c3a36850e90157da", "department": "Production", "gender": 2, "id": 61051, "job": "Executive Producer", "name": "Andrew Reimer"}, {"credit_id": "52fe44e8c3a368484e03da8d", "department": "Directing", "gender": 2, "id": 85563, "job": "Director", "name": "Brian Herzlinger"}, {"credit_id": "52fe44e8c3a368484e03da87", "department": "Directing", "gender": 2, "id": 94471, "job": "Director", "name": "Jon Gunn"}, {"credit_id": "52fe44e8c3a368484e03da97", "department": "Directing", "gender": 0, "id": 997560, "job": "Director", "name": "Brett Winn"}]'

In [36]:
all_df['Director_name']=all_df['crew'].apply(release_director_name)

In [37]:
all_df.head(2)

id                                     title  \
0  19995                                    Avatar   
1    285  Pirates of the Caribbean: At World's End   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]   

                                                crew     Director_name  
0  [{"credit_id": "52fe48009251416c750aca23", "de...   [James Cameron]  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  [Gore Verbinski]

In [38]:
all_df.drop('crew',axis=1,inplace=True)

In [39]:
all_df.sample(5)

id        title                                  genres  \
881    39437      Beloved                       [Drama, Thriller]   
3275    8293  Howards End                        [Drama, Romance]   
1586  227783  The Nut Job  [Animation, Comedy, Family, Adventure]   
2416   21765    Good Boy!       [Comedy, Family, Science Fiction]   
1509   22907       Takers        [Action, Crime, Drama, Thriller]   

                                               overview  \
881   After Paul D. finds his old slave friend Sethe...   
3275  Merchant Ivory’s adaptation of EM Forster’s cl...   
1586  Surly, a curmudgeon, independent squirrel is b...   
2416  An intergalactic dog pilot from Sirius (the do...   
1509  A seasoned team of bank robbers, including Gor...   

                                               keywords  \
881                                                  []   
3275  [based on novel, sister sister relationship, e...   
1586                                     [squirrel, 3d]   
2416                                                 []   
1509                                            [heist]   

                                                   cast          Director_name  
881       [Oprah Winfrey, Danny Glover, Thandie Newton]       [Jonathan Demme]  
3275  [Vanessa Redgrave, Helena Bonham Carter, Emma ...          [James Ivory]  
1586     [Will Arnett, Katherine Heigl, Brendan Fraser]     [Peter Lepeniotis]  
2416     [Molly Shannon, Liam Aiken, Matthew Broderick]  [John Robert Hoffman]  
1509     [Chris Brown, Hayden Christensen, Matt Dillon]      [John Luessenhop]

#### NOW,  WE HAVE TO REMOVE ALL THE SPACE BETWEEN EVERY WORD IN EVERY VARIABLE OF TAGS TO AVOID CONFUSION FOR THE MODEL i.e.
#### => we have to convert "Vishal Raj" a name to "VishalRaj"
##### AS TAGS WILL ALWAYS HAVE NO SPACE B/W THEM

In [40]:
all_df['genres']=all_df['genres'].apply(lambda i:[w.replace(" ","") for w in i])
all_df['keywords']=all_df['keywords'].apply(lambda i:[w.replace(" ","") for w in i])
all_df['cast']=all_df['cast'].apply(lambda i:[w.replace(" ","") for w in i])
all_df['Director_name']=all_df['Director_name'].apply(lambda i:[w.replace(" ","") for w in i])

In [41]:
all_df.sample(7)

id                                     title  \
1466   11876                  The Horseman on the Roof   
3605   11357  Halloween 4: The Return of Michael Myers   
4275  218500             The Ballad of Gregorio Cortez   
3377   13006                              Split Second   
297     1372                             Blood Diamond   
2623   72359                                   Peeples   
3825   29262                         The House of Sand   

                                                 genres  \
1466                   [War, Adventure, Drama, Romance]   
3605                                 [Horror, Thriller]   
4275                                          [Western]   
3377  [Thriller, Action, Horror, ScienceFiction, Crime]   
297                           [Drama, Thriller, Action]   
2623                                           [Comedy]   
3825                                            [Drama]   

                                               overview  \
1466  In a time of war and disease, a young officer ...   
3605  The legend of that creepy masked-man, Michael ...   
4275  The entire cause of the problem evolves from t...   
3377  In a futuristic London, the rising sea levels ...   
297   An ex-mercenary turned smuggler. A Mende fishe...   
2623  The story follows what happens when a child ps...   
3825  A woman is taken along with her mother in 1910...   

                                               keywords  \
1466         [italian, newlove, provence, horse, exile]   
3605  [sheriff, scissors, ambulance, halloween, twis...   
4275                                                 []   
3377            [flooding, futuristic, policedetective]   
297   [rebel, journalist, journalism, lossoffamily, ...   
2623              [duringcreditsstinger, womandirector]   
3825                            [brazil, desert, 1910s]   

                                                   cast         Director_name  
1466   [JulietteBinoche, OlivierMartinez, PierreArditi]  [Jean-PaulRappeneau]  
3605    [DonaldPleasence, EllieCornell, DanielleHarris]      [DwightH.Little]  
4275          [EdwardJamesOlmos, JamesGammon, TomBower]       [RobertM.Young]  
3377       [RutgerHauer, KimCattrall, MichaelJ.Pollard]          [TonyMaylam]  
297   [LeonardoDiCaprio, DjimonHounsou, JenniferConn...         [EdwardZwick]  
2623   [CraigRobinson, KerryWashington, DavidAlanGrier]     [TinaGordonChism]  
3825    [FernandaMontenegro, RuyGuerra, FernandaTorres]  [AndruchaWaddington]

In [42]:
all_df['overview']=all_df['overview'].apply(lambda v:v.split())
# Converting "Overview" variable into a list to combine with other columns as it isin string format
all_df.sample(7)

id                          title  \
867     242        The Godfather: Part III   
223    2789      The Chronicles of Riddick   
4190   4929                  Hang 'em High   
2182  22796                    Simon Birch   
325   57800     Ice Age: Continental Drift   
2363  38665  The Legend of the Lone Ranger   
2397  75638                     Red Lights   

                                      genres  \
867                 [Crime, Drama, Thriller]   
223                 [Action, ScienceFiction]   
4190                               [Western]   
2182                 [Comedy, Drama, Family]   
325   [Animation, Comedy, Adventure, Family]   
2363   [Action, Adventure, Romance, Western]   
2397                              [Thriller]   

                                               overview  \
867   [In, the, midst, of, trying, to, legitimize, h...   
223   [After, years, of, outrunning, ruthless, bount...   
4190  [Marshall, Jed, Cooper, survives, a, hanging,,...   
2182  [Simon, Birch, tells, the, story, of, Joe, and...   
325   [Manny,, Diego,, and, Sid, embark, upon, anoth...   
2363  [When, the, young, Texas, Ranger,, John, Reid,...   
2397  [Two, investigators, of, paranormal, hoaxes,, ...   

                                               keywords  \
867   [italy, christianity, newyork, assassination, ...   
223   [prison, dystopia, matteroflifeanddeath, outer...   
4190  [prison, judge, marshal, deathpenalty, oklahom...   
2182                                  [illegitimateson]   
325   [bluefootedbooby, prehistorictimes, meltingice...   
2363  [loneranger, wildbillhickok, buffalobill, thel...   
2397                  [paranormal, psychic, skepticism]   

                                                cast  \
867              [AlPacino, DianeKeaton, AndyGarcía]   
223            [VinDiesel, ThandieNewton, KarlUrban]   
4190         [ClintEastwood, IngerStevens, EdBegley]   
2182        [JimCarrey, AshleyJudd, IanMichaelSmith]   
325           [JohnLeguizamo, RayRomano, ChrisWedge]   
2363       [ChristopherLloyd, MattClark, JuaninClay]   
2397  [SigourneyWeaver, RobertDeNiro, CillianMurphy]   

                      Director_name  
867            [FrancisFordCoppola]  
223                    [DavidTwohy]  
4190                      [TedPost]  
2182            [MarkStevenJohnson]  
325   [SteveMartino, MikeThurmeier]  
2363              [WilliamA.Fraker]  
2397                [RodrigoCortés]

#### COMBINING EVERY COLUMN EXCEPT 'ID' AND 'TITLE' AS ALL_TAGS TO GET ALL THE TAGS RELATED TO A MOVIE IN A SINGLE COLUMN

In [43]:
all_df['all_Tags']=all_df['genres']+all_df['overview']+all_df['keywords']+all_df['cast']+all_df['Director_name']
all_df.drop(['genres','overview','keywords','cast','Director_name'],axis=1, inplace=True)

In [44]:
all_df.sample(5)

id                      title  \
2937   13937               Raising Cain   
3701    9336             Police Academy   
3472   25066        Tales from the Hood   
2568  232679  When the Game Stands Tall   
1137    9607          Super Mario Bros.   

                                               all_Tags  
2937  [Drama, Horror, Thriller, Crime, When, neighbo...  
3701  [Comedy, Crime, New, rules, enforced, by, the,...  
3472  [Crime, Horror, Thriller, A, strange, funeral,...  
2568  [Drama, A, young, coach, turns, a, losing, hig...  
1137  [Adventure, Comedy, Family, Fantasy, Mario, an...

In [45]:
all_df['all_Tags'][0]

['Action',
 'Adventure',
 'Fantasy',
 'ScienceFiction',
 'In',
 'the',
 '22nd',
 'century,',
 'a',
 'paraplegic',
 'Marine',
 'is',
 'dispatched',
 'to',
 'the',
 'moon',
 'Pandora',
 'on',
 'a',
 'unique',
 'mission,',
 'but',
 'becomes',
 'torn',
 'between',
 'following',
 'orders',
 'and',
 'protecting',
 'an',
 'alien',
 'civilization.',
 'cultureclash',
 'future',
 'spacewar',
 'spacecolony',
 'society',
 'spacetravel',
 'futuristic',
 'romance',
 'space',
 'alien',
 'tribe',
 'alienplanet',
 'cgi',
 'marine',
 'soldier',
 'battle',
 'loveaffair',
 'antiwar',
 'powerrelations',
 'mindandsoul',
 '3d',
 'SamWorthington',
 'ZoeSaldana',
 'SigourneyWeaver',
 'JamesCameron']

#### CONVERTING LIST TO STRING

In [46]:
all_df['all_Tags']=all_df['all_Tags'].apply(lambda x:" ".join(x))

In [47]:
all_df.sample(5)

id                               title  \
2015   11156                  Coco Before Chanel   
2774  107811                           21 & Over   
1105   68734                                Argo   
981      421  The Life Aquatic with Steve Zissou   
1596    4347                           Atonement   

                                               all_Tags  
2015  Drama History 2009 biopic about the early life...  
2774  Comedy Brilliant student Jeff Chang has the mo...  
1105  Drama Thriller As the Iranian revolution reach...  
981   Adventure Comedy Drama Wes Anderson’s incisive...  
1596  Drama Romance As a 13-year-old, fledgling writ...

In [48]:
all_df['all_Tags'][5]

"Fantasy Action Adventure The seemingly invincible Spider-Man goes up against an all-new crop of villain – including the shape-shifting Sandman. While Spider-Man’s superpowers are altered by an alien organism, his alter ego, Peter Parker, deals with nemesis Eddie Brock and also gets caught up in a love triangle. dualidentity amnesia sandstorm loveofone'slife forgiveness spider wretch deathofafriend egomania sand narcism hostility marvelcomic sequel superhero revenge TobeyMaguire KirstenDunst JamesFranco SamRaimi"

In [49]:
all_df['all_Tags']=all_df['all_Tags'].apply(lambda x:x.lower())

In [50]:
all_df

id                                     title  \
0      19995                                    Avatar   
1        285  Pirates of the Caribbean: At World's End   
2     206647                                   Spectre   
3      49026                     The Dark Knight Rises   
4      49529                               John Carter   
...      ...                                       ...   
4804    9367                               El Mariachi   
4805   72766                                 Newlyweds   
4806  231617                 Signed, Sealed, Delivered   
4807  126186                          Shanghai Calling   
4808   25975                         My Date with Drew   

                                               all_Tags  
0     action adventure fantasy sciencefiction in the...  
1     adventure fantasy action captain barbossa, lon...  
2     action adventure crime a cryptic message from ...  
3     action crime drama thriller following the deat...  
4     action adventure sciencefiction john carter is...  
...                                                 ...  
4804  action crime thriller el mariachi just wants t...  
4805  comedy romance a newlywed couple's honeymoon i...  
4806  comedy drama romance tvmovie "signed, sealed, ...  
4807  when ambitious new york attorney sam is sent t...  
4808  documentary ever since the second grade when h...  

[4806 rows x 3 columns]

In [51]:
# Removing HTML tags in "all_Tags" variable
from bs4 import BeautifulSoup

In [52]:
text_html='<h1><p>I am Akshay Kumar</p>'
s=BeautifulSoup(text_html,'html.parser')
print(s.get_text())

I am Akshay Kumar


In [53]:
def remove_htmlTags(text):
    s=BeautifulSoup(text,'html.parser')
    return s.get_text()
all_df['all_Tags']=all_df['all_Tags'].apply(remove_htmlTags)
all_df.sample(6)

C:\Users\aksha\AppData\Local\Temp\ipykernel_18752\3508911121.py:2: MarkupResemblesLocatorWarning: The input looks more like a filename than markup. You may want to open this file and pass the filehandle into Beautiful Soup.
  s=BeautifulSoup(text,'html.parser')


id               title  \
3772   9809  Let's Go to Prison   
142   11619        Flushed Away   
1413  10066        House of Wax   
1411    423         The Pianist   
2699  12637  New York, New York   
3406  78381               Twixt   

                                               all_Tags  
3772  comedy crime when a career criminal's plan for...  
142   adventure animation comedy family london high-...  
1413  horror a group of unwitting teens are stranded...  
1411  drama war the pianist is a film adapted from t...  
2699  drama music romance an egotistical saxophone p...  
3406  horror thriller a writer with a declining care...

In [54]:
# Removing punctuations and numbers
import string
puncs_nums=string.punctuation+"0123456789’–"
puncs_nums

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~0123456789’–'

In [55]:
test=all_df.all_Tags.iloc[223]
test

"action sciencefiction after years of outrunning ruthless bounty hunters, escaped convict riddick suddenly finds himself caught between opposing forces in a fight for the future of the human race. now, waging incredible battles on fantastic and deadly worlds, this lone, reluctant hero will emerge as humanity's champion - and the last hope for a universe on the edge of annihilation. prison dystopia matteroflifeanddeath outerspace intergalactictravel vindiesel thandienewton karlurban davidtwohy"

In [56]:
def remove_puncs_nums(text):
    for i in puncs_nums:
        text=text.replace(i,"")
    return text

In [57]:
remove_puncs_nums(test)

'action sciencefiction after years of outrunning ruthless bounty hunters escaped convict riddick suddenly finds himself caught between opposing forces in a fight for the future of the human race now waging incredible battles on fantastic and deadly worlds this lone reluctant hero will emerge as humanitys champion  and the last hope for a universe on the edge of annihilation prison dystopia matteroflifeanddeath outerspace intergalactictravel vindiesel thandienewton karlurban davidtwohy'

In [58]:
all_df['all_Tags'].iloc[5]

"fantasy action adventure the seemingly invincible spider-man goes up against an all-new crop of villain – including the shape-shifting sandman. while spider-man’s superpowers are altered by an alien organism, his alter ego, peter parker, deals with nemesis eddie brock and also gets caught up in a love triangle. dualidentity amnesia sandstorm loveofone'slife forgiveness spider wretch deathofafriend egomania sand narcism hostility marvelcomic sequel superhero revenge tobeymaguire kirstendunst jamesfranco samraimi"

In [59]:
all_df['all_Tags']=all_df['all_Tags'].apply(remove_puncs_nums)
all_df.sample(4)

id              title  \
2484  25166      Drowning Mona   
3004  10426  Friday After Next   
3442  62255            Tracker   
1395  22949           Old Dogs   

                                               all_Tags  
2484  comedy crime mystery the recently deceased mon...  
3004  comedy craig and day day have finally moved ou...  
3442  action drama thriller adventure an exboer war ...  
1395  comedy family charlie and dan have been best f...

In [60]:
all_df['all_Tags'].iloc[5]

'fantasy action adventure the seemingly invincible spiderman goes up against an allnew crop of villain  including the shapeshifting sandman while spidermans superpowers are altered by an alien organism his alter ego peter parker deals with nemesis eddie brock and also gets caught up in a love triangle dualidentity amnesia sandstorm loveofoneslife forgiveness spider wretch deathofafriend egomania sand narcism hostility marvelcomic sequel superhero revenge tobeymaguire kirstendunst jamesfranco samraimi'

In [61]:
import nltk
from nltk.stem import WordNetLemmatizer
nltk.download('wordnet')
wnl=WordNetLemmatizer()

[nltk_data] Downloading package wordnet to C:\Users\aksha/nltk_data...


In [62]:
test=all_df.all_Tags.iloc[39]
test

'adventure action sciencefiction sam flynn the techsavvy and daring son of kevin flynn investigates his fathers disappearance and is pulled into the grid with the help of a mysterious program named quorra sam quests to stop evil dictator clu from crossing into the real world artificialintelligence secretidentity computerprogram dystopia computer deception duel motorcycle neonlight autocracy garretthedlund jeffbridges oliviawilde josephkosinski'

In [63]:
lemmatized_test = [wnl.lemmatize(word, pos="v") for word in test.split()]

lemmatized_test_str = " ".join(lemmatized_test)

lemmatized_test_str

'adventure action sciencefiction sam flynn the techsavvy and dare son of kevin flynn investigate his father disappearance and be pull into the grid with the help of a mysterious program name quorra sam quest to stop evil dictator clu from cross into the real world artificialintelligence secretidentity computerprogram dystopia computer deception duel motorcycle neonlight autocracy garretthedlund jeffbridges oliviawilde josephkosinski'

In [64]:
def lemmatizer(text):
    lemmatized_test=[wnl.lemmatize(word, pos="v") for word in text.split()]
    lemmatized_test_str=" ".join(lemmatized_test)
    return lemmatized_test_str

In [65]:
all_df['all_Tags']=all_df['all_Tags'].apply(lemmatizer)
all_df.sample(4)

id                    title  \
517    11026  Exorcist: The Beginning   
2523   45269        The King's Speech   
2886  179144         The Great Beauty   
2762   10691          Crazy/Beautiful   

                                               all_Tags  
517   horror mystery thriller have live through trau...  
2523  drama history the kings speech tell the story ...  
2886  comedy drama jep gambardella have seduce his w...  
2762  drama romance at pacific palisade high a poor ...

In [66]:
all_df.shape

(4806, 3)

In [67]:
# Vector Representation

In [68]:
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(max_features=5000,stop_words='english')

In [69]:
all_vectors=cv.fit_transform(all_df['all_Tags']).toarray()
all_vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [70]:
len(cv.get_feature_names_out())

5000

In [71]:
# These are all the features used
for i in cv.get_feature_names_out():
    print(i)

aaron
aaroneckhart
aarontaylorjohnson
abandon
abduct
abigailbreslin
abilities
ability
able
aboard
abuse
abusive
academic
academy
accept
access
accident
accidental
accidentally
acclaim
accompany
accomplish
account
accountant
accuse
ace
achieve
act
action
actionhero
actionpacked
activist
activities
activity
actor
actors
actress
actual
actually
ad
adam
adammckay
adams
adamsandler
adamshankman
adapt
adaptation
add
addict
addiction
adjust
admire
admit
adolescence
adolescent
adopt
adoption
adrienbrody
adult
adultanimation
adultery
adulthood
adults
advance
advantage
adventure
advertise
advice
affair
affairs
affect
affection
affections
affluent
afghanistan
afraid
africa
african
africanamerican
aftercreditsstinger
afterlife
aftermath
age
agedifference
agency
agenda
agent
agents
aggressive
ago
agree
ahead
aid
aidanquinn
ail
aim
air
airplane
airplanecrash
airport
aka
al
alabama
alan
alanarkin
alaska
albert
albertbrooks
alcatraz
alcohol
alcoholabuse
alcoholic
alcoholism
alecbaldwin
alejandrogonzál

In [72]:
# Generating similarity scores using cosine_similarity for getting nearest vectors (i.e. recommended movies) for given input (i.e. movie)

In [73]:
from sklearn.metrics.pairwise import cosine_similarity

In [74]:
cosine_similarity(all_vectors).shape

(4806, 4806)

In [75]:
all_similarities=cosine_similarity(all_vectors)

In [76]:
all_similarities[1]

array([0.08858079, 1.        , 0.06253054, ..., 0.02514977, 0.        ,
       0.        ])

In [77]:
# RETURNS FIRST FIVE NEAREST VECTORS INDICES WHEN GIVEN A INPUT VECTOR INDEX, hERE, INPUT VECTOR INDEX IS:1
sorted(list(enumerate(all_similarities[1])),reverse=True,key=lambda c:c[1])[1:6]

[(12, 0.4412487516463849),
 (199, 0.2988812888320308),
 (17, 0.27371875400769585),
 (216, 0.21552636243212991),
 (2135, 0.20526320231631418)]

In [78]:
def recommend(movie):
    movie_index=all_df[all_df['title']==movie].index[0]
    distances=all_similarities[movie_index]
    movies_list=sorted(list(enumerate(distances)),reverse=True,key=lambda c:c[1])[1:6]
    print('\nRecommended movies based on the input movie you entered:\n')
    for i in movies_list:
        print(all_df.iloc[i[0]].title)

In [79]:
all_movies_titles_available=all_df['title'].values
for available_movie in sorted(all_movies_titles_available):
    print(available_movie)

#Horror
(500) Days of Summer
10 Cloverfield Lane
10 Days in a Madhouse
10 Things I Hate About You
102 Dalmatians
10th & Wolf
11:14
12 Angry Men
12 Rounds
12 Years a Slave
127 Hours
13 Going on 30
13 Hours: The Secret Soldiers of Benghazi
1408
15 Minutes
16 Blocks
16 to Life
17 Again
1776
1911
1941
1982
2 Fast 2 Furious
2 Guns
20 Dates
20 Feet from Stardom
20,000 Leagues Under the Sea
200 Cigarettes
2001: A Space Odyssey
2012
2016: Obama's America
2046
21
21 & Over
21 Grams
21 Jump Street
22 Jump Street
24 7: Twenty Four Seven
25th Hour
27 Dresses
28 Days
28 Days Later
28 Weeks Later
2:13
3 Backyards
3 Days to Kill
3 Ninjas Kick Back
3 Strikes
30 Days of Night
30 Minutes or Less
30 Nights of Paranormal Activity With the Devil Inside the Girl With the Dragon Tattoo
300
3000 Miles to Graceland
300: Rise of an Empire
3:10 to Yuma
4 Months, 3 Weeks and 2 Days
40 Days and 40 Nights
42
42nd Street
47 Ronin
5 Days of War
50 First Dates
50/50
51 Birch Street
54
55 Days at Peking
8 Days
8 Heads 

In [90]:
try:
    movie_name=input('Please enter the movie you like from the above listed movies, so that we can recommend movies that are similar to the given movie:\n')
    recommend(movie_name)
except:
    print(f"\n{movie_name} movie is not available in this project.\nPlease provide another movie name for recommendations.")


Recommended movies based on the input movie you entered:

Batman
Batman & Robin
Batman Begins
Batman Forever
The Dark Knight Rises


In [81]:
len(all_movies_titles_available)

4806

In [82]:
dff=pd.DataFrame(all_movies_titles_available)
dff.to_html('all_movies_list',index=False)

In [83]:
import pickle

In [91]:
pickle.dump(all_df.to_dict(),open('df_dict.pkl','wb'))

In [92]:
pickle.dump(all_similarities,open('all_similarities.pkl','wb'))